src/Bedrock.Framework.Experimental/Protocols/Kafka/PayloadWriter.cs

```diff
@@ -0,0 +1,235 @@
+        {
+            this.Context = settings;
+
+            // Hook up the parent's pipe reader to this writer, and vice-versa.
+            this.ToParent = PipeWriter.Create(this.Context.Pipe.Reader.AsStream());
+            this.FromChild = PipeReader.Create(this.Context.Pipe.Writer.AsStream());
+
+            this.CurrentWriter = this.Context.Pipe.Writer;
+        }
+
+        /// <summary>
+        /// Creates a root instance of the <see cref="PayloadWriter"/> struct.
+        /// </summary>
+        /// <param name="isBigEndian">Whether or not to write bytes as big endian. Defaults to true.</param>
+        public PayloadWriter(bool isBigEndian)
+        {
+            var pipe = new Pipe();
+            this.ToParent = pipe.Writer;
+            this.FromChild = pipe.Reader;
+
+            this.Context = new PayloadWriterContext(
+                isBigEndian,
+                pipe);
+
+            // Root writer, so route writer to itself.
+            this.CurrentWriter = this.ToParent;
+        }
+
+        /// <summary>
+        /// Sets the location in a payload where a size will be calculated.
+        /// Calculated size is between <see cref="StartCalculatingSize(string)"/>
+        /// and a call to <see cref="EndSizeCalculation(string)"/> with the same name.
+        /// </summary>
+        /// <param name="name">The distinct name of a size calculation.</param>
+        /// <returns>The <see cref="PayloadWriter"/>.</returns>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public PayloadWriter StartCalculatingSize(string name)
+        {
+            if (this.Context.SizeCalculations.ContainsKey(name))
+            {
+                throw new ArgumentException($"Unable to add another size calculation called: {name}", nameof(name));
+            }
+
+            var memory = this.CurrentWriter
+                .GetMemory(sizeof(int))
+                .Slice(0, sizeof(int));
+
+            // Size calculation is not inclusive of the size value itself, + sizeof(int) starts
+            // the calculation _after_ where the size value would be.
+            this.Context.SizeCalculations[name] = (this.Context.BytesWritten + sizeof(int), memory);
+
+            this.Context.Advance(sizeof(int));
+
+            return this;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public PayloadWriter EndSizeCalculation(string name)
+        {
+            if (!this.Context.SizeCalculations.TryGetValue(name, out var calculation))
+            {
+                throw new ArgumentException($"Size calculation for {name} not found", nameof(name));
+            }
+
+            this.Context.SizeCalculations.Remove(name);
+            var currentPosition = this.Context.BytesWritten;
+            var size = (int)(currentPosition - calculation.position);
+
+            var span = calculation.memory.Span;
+
+            if (this.Context.IsBigEndian)
```

Comment:
This pattern is repeated for every call.
The StrategyPayloadWriter uses an interface to remove this call...but, isn't any faster.